### PR TITLE
[CB] Slice logits inside the model

### PR DESCRIPTION
**Pull request by remi-or**
# Summary

This PR leverages an **awesome** kwarg in transformers: `logits_to_keep`. Simply, this drop the tokens for which we won't use the next token predictions before we enter the `lm_head`. Reduces the memory footprints of large prefill batches, increase their throughput, makes everyone happy.

## Performance ✅ 

label | main acc | new PR acc | main (tok/s) | new PR (tok/s) | diff
-- | -- | -- | -- | -- | --
gsm8k_default | 0.822 | 0.822 | 2460.91 | 2477.15 | +0.66%
gsm8k_sampling | 0.787 | 0.784 | 1952.33 | 1958.65 | +0.32%
gsm8k_compile | 0.821 | 0.821 | 2461.93 | 2473.02 | +0.45%
gsm8k_no_fast_decode | 0.822 | 0.822 | 2350.41 | 2355.21 | +0.20%
gsm8k_bare_bones | 0.821 | 0.821 | 1845.35 | 1845.52 | +0.01%
ifeval_default | 0.445 | 0.447 | 9007.60 | 9015.87 | +0.09%
few_blocks | - | - | 666.92 | 664.97 | -0.29%
multi_return_seq | - | - | 1610.03 | 1627.67 | +1.10%
rollouts_1024 | - | - | 3546.62 | 3550.29 | +0.10%
rollouts_2048 | - | - | 3371.13 | 3373.68 | +0.08%
rollouts_4096 | - | - | 2973.00 | 2974.99 | +0.07%
rollouts_8192 | - | - | 2361.10 | 2362.95 | +0.08%
rollouts_16384 | - | - | 1549.11 | 1549.40 | +0.02%

No performance degradation.

## Tests ✅ 

All tests pass

## AI Reviewed ✅ 

Done and addressed

### CODE REVIEW

**Review by ArthurZucker — APPROVED** [review]
Nice
